src/workerd/api/basics.h

```diff
@@ -224,7 +224,33 @@ class EventTarget: public jsg::Object {
 public:
 
   // RAII-style listener that can be attached to an EventTarget.
-  class NativeHandler;
+  class NativeHandler {
+  public:
+    using Signature = void(jsg::Ref<Event>);
+    NativeHandler(jsg::Lock& js, jsg::Ref<EventTarget> target, kj::String type,
+        jsg::Function<Signature> func, bool once = false);
+    ~NativeHandler() noexcept(false);
+    KJ_DISALLOW_COPY_AND_MOVE(NativeHandler);
+
+    void operator()(jsg::Lock&js, jsg::Ref<Event> event);
+
+    uint hashCode() const;
+
+    void visitForGc(jsg::GcVisitor& visitor);
+  private:
+    void detach();
+
+    kj::String type;
+    struct State {
+      jsg::Ref<EventTarget> target;
+      jsg::Function<Signature> func;
+    };
+    kj::Maybe<State> state;
```

Comment:
maybe one-liner comment about when it is none? E.g. // is never empty during operations, gets cleared in detach()